File: ui/src/checkers/secrets/required_present.ts
/**
 * Verifies all required Kubernetes secrets are present and non-empty.
 *
 * Looks at the secrets/required_present target (boolean or string).
 * This includes zAuth keys (brig+nginz), TURN secret, SMTP password,
 * RabbitMQ credentials, PG password, MinIO keys, and the TLS wildcard
 * cert+key. Without them, services crash on startup.
 */

// Ours
import { BaseChecker, type CheckResult } from '../base_checker'
import type { DataLookup } from '../data_lookup'

export class RequiredPresentChecker extends BaseChecker {
    readonly path: string = 'secrets/required_present'
    readonly name: string = 'All required K8s Secrets present and non-empty'
    readonly category: string = 'Secrets / Credentials'
    readonly interest = 'Setup' as const
    readonly explanation: string = 'Verifies that all required **Kubernetes secrets** (`zAuth` keys, `TURN` secret, SMTP password, RabbitMQ credentials, TLS certs) exist and are non-empty. Missing secrets cause services to **crash on startup**.'

    check(data: DataLookup): CheckResult {
        const point = data.get('secrets/required_present')

        // Target data was not collected
        if (!point) {
            return {
                status: 'gather_failure',
                status_reason: 'Target data for `secrets/required_present` was not collected by the gatherer.',
                fix_hint: '1. Verify the gatherer has **kubectl access** to the cluster\n2. Check that the gatherer can run `kubectl get secrets -n wire`\n3. Review the gatherer logs for permission errors or timeouts',
                recommendation: 'All required K8s Secrets present and non-empty data not collected.',
            }
        }

        // Data point exists but value is null — gatherer could not reach the cluster
        if (point.value === null) {
            return {
                status: 'gather_failure',
                status_reason: 'Secrets presence data was collected but contained no value.',
                recommendation: point.metadata?.error ?? 'Required secrets target ran but returned no result.',
                raw_output: point.raw_output,
            }
        }

        const val: string | boolean = point.value as string | boolean

        // Extract specific missing secrets from health_info for status_reason
        const missing_issues: string = this._extract_issues(point)

        // String value non-empty means all secrets are present
        if (typeof val === 'string') {
            if (val.length > 0) {
                return {
                    status: 'healthy',
                    status_reason: 'All required K8s secrets are **present** and non-empty.',
                    display_value: val,
                    raw_output: point.raw_output,
                }
            }

            return {
                status: 'unhealthy',
                status_reason: missing_issues
                    ? 'Required K8s secrets are **missing**: {{missing_issues}}.'
                    : 'One or more required K8s secrets are **missing or empty**.',
                fix_hint: missing_issues
                    ? '1. List the current secrets in the Wire namespace:\n   ```\n   kubectl get secrets -n wire\n   ```\n2. Create the missing secrets ({{missing_issues}}) using the Wire installation guide\n3. For TLS secrets, verify the certificate and key files are valid before creating the secret:\n   ```\n   openssl x509 -in tls.crt -noout -text\n   ```\n4. Restart affected pods after creating the secrets: `kubectl rollout restart deployment -n wire`'
                    : '1. List the current secrets in the Wire namespace:\n   ```\n   kubectl get secrets -n wire\n   ```\n2. Compare against the required secrets list: `zAuth` keys, `TURN` secret, SMTP password, RabbitMQ credentials, TLS certs\n3. Create any missing secrets following the Wire installation guide\n4. Restart affected pods after creating the secrets: `kubectl rollout restart deployment -n wire`',
                recommendation: this._build_missing_recommendation(point),
                display_value: val,
                raw_output: point.raw_output,
                template_data: { missing_issues },
            }
        }

        // Boolean true means all secrets are present
        if (val === true) {
            return {
                status: 'healthy',
                status_reason: 'All required K8s secrets are **present** and non-empty.',
                display_value: val,
                raw_output: point.raw_output,
            }
        }

        // Boolean false some secrets are missing
        return {
            status: 'unhealthy',
            status_reason: missing_issues
                ? 'Required K8s secrets are **missing**: {{missing_issues}}.'
                : 'One or more required K8s secrets are **missing or empty**.',
            fix_hint: missing_issues
                ? '1. List the current secrets in the Wire namespace:\n   ```\n   kubectl get secrets -n wire\n   ```\n2. Create the missing secrets ({{missing_issues}}) using the Wire installation guide\n3. For TLS secrets, verify the certificate and key files are valid before creating the secret:\n   ```\n   openssl x509 -in tls.crt -noout -text\n   ```\n4. Restart affected pods after creating the secrets: `kubectl rollout restart deployment -n wire`'
                : '1. List the current secrets in the Wire namespace:\n   ```\n   kubectl get secrets -n wire\n   ```\n2. Compare against the required secrets list: `zAuth` keys, `TURN` secret, SMTP password, RabbitMQ credentials, TLS certs\n3. Create any missing secrets following the Wire installation guide\n4. Restart affected pods after creating the secrets: `kubectl rollout restart deployment -n wire`',
            recommendation: this._build_missing_recommendation(point),
            display_value: val,
            raw_output: point.raw_output,
            template_data: { missing_issues },
        }
    }

    /** Extract the comma-separated issues list from health_info, or empty string if not available. */
    private _extract_issues(point: { metadata?: { health_info?: string } }): string {
        const health_info: string = point.metadata?.health_info ?? ''
        const issues_match: RegExpMatchArray | null = health_info.match(/Issues:\s*(.+)/)
        return issues_match?.[1] ?? ''
    }

    /** Extract specific missing secrets from the collector's health_info and build a targeted recommendation. */
    private _build_missing_recommendation(point: { metadata?: { health_info?: string } }): string {
        const health_info: string = point.metadata?.health_info ?? ''

        // format: «Issues: brig-secrets, nginz-secrets, brig-turn, wire-server-tls»
        const issues_match: RegExpMatchArray | null = health_info.match(/Issues:\s*(.+)/)
        const specific_issues: string = issues_match?.[1] ?? ''

        if (specific_issues) {
            return `Missing or incomplete K8s secrets: ${specific_issues}. Services will crash on startup without these.`
        }

        return 'Required K8s secrets are missing. Services crash on startup without them.'
    }
}

export default RequiredPresentChecker

File: src/targets/secrets/required_present.py
"""Checks that all required Kubernetes Secrets exist and have data.

Missing a secret and the service crashes on startup. We check zAuth keys
(brig + nginz), TURN secret, SMTP password, RabbitMQ creds, PG password,
MinIO keys, and TLS cert+key.
"""

from __future__ import annotations

# External
from typing import Any

# Ours
from src.lib.base_target import BaseTarget


# Secrets that must exist for Wire to work.
# Each entry: (list of possible secret names, description, list of required keys).
# We accept ANY of the listed names different deploy types (helm, WIAB, manual)
# use different naming conventions.
# When required_keys is non-empty, we verify those keys exist in the matching
# secret (e.g. TURN secret might be a key inside brig secret rather than standalone).
_REQUIRED_SECRETS: list[tuple[list[str], str, list[str]]] = [
    (
        ["brig-secrets", "brig"],
        "Brig service secrets (zAuth keys)",
        [],
    ),
    (
        ["nginz-secrets", "nginz"],
        "Nginz gateway secrets (zAuth public key)",
        [],
    ),
    (
        ["brig-turn", "brig"],
        "TURN server secret",
        ["turn-secret.txt"],
    ),
    (
        ["wire-server-tls"],
        "TLS wildcard certificate",
        ["tls.crt", "tls.key"],
    ),
]

# Additional secrets to check if they exist (may have different names per deployment)
_OPTIONAL_SECRET_PATTERNS: list[str] = [
    "smtp",
    "rabbitmq",
    "minio",
    "postgresql",
    "postgres",
]


class RequiredSecretsPresent(BaseTarget):
    """Checks that required Kubernetes Secrets exist and have data.

    Queries secrets in the Wire namespace and verifies each required secret
    exists and contains non-empty data.
    """

    @property
    def description(self) -> str:
        """Human-readable description of what this target checks."""
        return "All required Kubernetes Secrets present and non-empty"

    @property
    def explanation(self) -> str:
        """Why this check exists and what determines healthy vs unhealthy."""
        return (
            "Missing secrets just crash services on startup. We check zAuth keys, TURN secret, "
            "TLS cert, and optional SMTP/RabbitMQ/MinIO/PostgreSQL secrets."
        )

    @property
    def unit(self) -> str:
        """Unit of measurement empty because result is boolean."""
        return ""

    def collect(self) -> bool:
        """Check that required secrets exist and have data.

        Returns:
            True if all required secrets are present, False otherwise.
        """
        self.terminal.step("Checking required Kubernetes secrets...")

        # Get all secrets in the namespace
        cmd_result, data = self.run_kubectl("secrets")

        if data is None:
            raise RuntimeError("Failed to query secrets from kubectl")

        items: list[dict[str, Any]] = data.get("items", [])

        # Build a map of secret names to their data keys
        secret_map: dict[str, set[str]] = {}
        for item in items:
            name: str = item.get("metadata", {}).get("name", "")
            data_keys: dict = item.get("data") or {}
            secret_map[name] = set(data_keys.keys())

        # Check required secrets accept any of the alternative names
        tls_desc: str = "TLS wildcard certificate"

        # Track the exact string for TLS in empty list so bypass can remove it by exact match
        # (entry is «wire-server-tls (empty)», not «... (TLS wildcard certificate)»,
        # so substring search would miss it)
        tls_empty_entry: str | None = None

        # Track exact string for TLS wrong-key case so bypass can remove by exact match
        # (entry like «wire-server-tls (missing keys: tls.crt, tls.key)» doesn't contain
        # tls_desc so substring filter would miss it)
        tls_missing_entry: str | None = None

        present: list[str] = []
        missing: list[str] = []
        empty: list[str] = []

        for possible_names, desc, required_keys in _REQUIRED_SECRETS:
            found_name: str | None = None
            for candidate in possible_names:
                if candidate in secret_map:
                    found_name = candidate
                    break

            if found_name is None:
                # None of the names matched
                missing.append(f"{possible_names[0]} ({desc})")
                continue

            keys_in_secret: set[str] = secret_map[found_name]

            if len(keys_in_secret) == 0:
                empty_entry: str = f"{found_name} (empty)"
                empty.append(empty_entry)
                # Remember the exact string so the TLS bypass can remove it by exact match
                if desc == tls_desc:
                    tls_empty_entry = empty_entry
                continue

            # If specific keys are required, verify they exist in the secret
            if required_keys:
                missing_keys: list[str] = [
                    k for k in required_keys if k not in keys_in_secret
                ]
                if missing_keys:
                    missing_entry: str = f"{found_name} (missing keys: {', '.join(missing_keys)})"
                    missing.append(missing_entry)
                    # Remember exact string so TLS bypass can remove it entry contains
                    # «missing keys: …» not tls_desc, so substring filter would miss it
                    if desc == tls_desc:
                        tls_missing_entry = missing_entry
                    continue

            present.append(found_name)

        # Also search for TLS cert/key in any secret (WIAB uses weird auto-generated names).
        # Check if wire-server-tls was added to present. Using any(...) on all present names
        # would be wrong an unrelated secret might have tls.crt/tls.key.
        tls_found_in_required: bool = "wire-server-tls" in present
        if not tls_found_in_required:
            # Check if any secret has both tls.crt and tls.key
            for name, keys in secret_map.items():
                if "tls.crt" in keys and "tls.key" in keys:
                    # Found a TLS secret remove any missing/empty entries for TLS check.
                    #
                    # Two formats can appear in missing:
                    #   1. «wire-server-tls (TLS wildcard certificate)» secret absent
                    #   2. «wire-server-tls (missing keys: tls.crt, tls.key)» exists but
                    #      incomplete (doesn't contain tls_desc, so need exact match)
                    missing = [m for m in missing if tls_desc not in m
                               and (tls_missing_entry is None or m != tls_missing_entry)]
                    # Filter by exact match entry is «wire-server-tls (empty)» not
                    # «... (TLS wildcard certificate)» so substring search would miss it
                    empty = [e for e in empty if tls_empty_entry is None or e != tls_empty_entry]
                    present.append(f"{name} (TLS)")
                    break

        # Check optional patterns just report if they exist
        optional_found: int = 0
        for pattern in _OPTIONAL_SECRET_PATTERNS:
            for name in secret_map:
                if pattern in name.lower():
                    optional_found += 1
                    break

        all_ok: bool = len(missing) == 0 and len(empty) == 0

        if all_ok:
            self._health_info = (
                f"All {len(_REQUIRED_SECRETS)} required secrets present, "
                f"{optional_found} optional patterns matched"
            )
        else:
            issues: list[str] = missing + empty
            self._health_info = f"Issues: {', '.join(issues)}"

        return all_ok
